vms/avm/service_test.go

```diff
@@ -18,9 +18,11 @@ import (
 func setup(t *testing.T) ([]byte, *VM, *Service) {
 	genesisBytes := BuildGenesisTest(t)
 
-	// TODO: Could this initialization be replaced by a call to GenesisVM()
 	ctx.Lock.Lock()
 
+	// This is VM initilialzation is very silimar to that done by GenesisVM().
+	// However that function also sets vm.batchTimeout = 0, which causes the
+	// tests in this module to deadlock.
```

Comment:
Sorry. I pressed the wrong button, and our previous thread is now MIA. You're right that adding `vm.batchTImeout = 0` here doesn't change the behaviour. However, when I replace the vm initialization in this function with `genesisBytes, _, vm := GenesisVM(t)`, then I get the following happens

```
~/src/ava-labs/gecko vms-avm-service-test(+193/-29)+*± scripts/build_test.sh
# github.com/ava-labs/go-ethereum/vendor/gopkg.in/olebedev/go-duktape.v3
In file included from _cgo_export.c:4:
debugger.go:22:13: warning: ‘_duk_debugger_attach’ defined but not used [-Wunused-function]
ok      _/home/alex/src/ava-labs/gecko/api      1.011s  coverage: 63.6% of statements
?       _/home/alex/src/ava-labs/gecko/api/admin        [no test files]
?       _/home/alex/src/ava-labs/gecko/api/ipcs [no test files]
...
?       _/home/alex/src/ava-labs/gecko/utils/units      [no test files]
ok      _/home/alex/src/ava-labs/gecko/utils/wrappers   1.016s  coverage: 61.2% of statements
?       _/home/alex/src/ava-labs/gecko/vms      [no test files]
coverage: 46.1% of statements
panic: test timed out after 10m0s

goroutine 132 [running]:
testing.(*M).startAlarm.func1()
        /usr/local/go/src/testing/testing.go:1377 +0x11c
created by time.goFunc
        /usr/local/go/src/time/sleep.go:168 +0x52

goroutine 1 [chan receive, 9 minutes]:
testing.(*T).Run(0xc000186000, 0xc2e39e, 0x17, 0xc48a50, 0x1)
        /usr/local/go/src/testing/testing.go:961 +0x68a
testing.runTests.func1(0xc000186000)
        /usr/local/go/src/testing/testing.go:1202 +0xa7
testing.tRunner(0xc000186000, 0xc0000c1cd8)
        /usr/local/go/src/testing/testing.go:909 +0x19a
testing.runTests(0xc00000e220, 0x107d220, 0x48, 0x48, 0x0)
        /usr/local/go/src/testing/testing.go:1200 +0x522
testing.(*M).Run(0xc000182000, 0x0)
        /usr/local/go/src/testing/testing.go:1117 +0x300
main.main()
        _testmain.go:268 +0x338

...

goroutine 59 [select, 9 minutes]:
github.com/ava-labs/gecko/utils/timer.(*Timer).Dispatch(0xc000232720)
        /home/alex/.local/go/src/github.com/ava-labs/gecko/utils/timer/timer.go:90 +0x289
github.com/ava-labs/gecko/utils/logging.NoLog.RecoverAndPanic(...)
        /home/alex/.local/go/src/github.com/ava-labs/gecko/utils/logging/test_log.go:53
created by _/home/alex/src/ava-labs/gecko/vms/avm.(*VM).Initialize
        /home/alex/src/ava-labs/gecko/vms/avm/vm.go:193 +0x1a84

goroutine 72 [semacquire, 9 minutes]:
sync.runtime_SemacquireMutex(0xc00015a03c, 0x900000000, 0x1)
        /usr/local/go/src/runtime/sema.go:71 +0x47
sync.(*Mutex).lockSlow(0xc00015a038)
        /usr/local/go/src/sync/mutex.go:138 +0x1c1
sync.(*Mutex).Lock(0xc00015a038)
        /usr/local/go/src/sync/mutex.go:81 +0x7d
sync.(*RWMutex).Lock(0xc00015a038)
        /usr/local/go/src/sync/rwmutex.go:98 +0x4a
_/home/alex/src/ava-labs/gecko/vms/avm.(*VM).Initialize.func1()
        /home/alex/src/ava-labs/gecko/vms/avm/vm.go:188 +0x7c
github.com/ava-labs/gecko/utils/timer.(*Timer).Dispatch(0xc000021dd0)
        /home/alex/.local/go/src/github.com/ava-labs/gecko/utils/timer/timer.go:83 +0x195
github.com/ava-labs/gecko/utils/logging.NoLog.RecoverAndPanic(...)
        /home/alex/.local/go/src/github.com/ava-labs/gecko/utils/logging/test_log.go:53
created by _/home/alex/src/ava-labs/gecko/vms/avm.(*VM).Initialize
        /home/alex/src/ava-labs/gecko/vms/avm/vm.go:193 +0x1a84
FAIL    _/home/alex/src/ava-labs/gecko/vms/avm  600.104s
ok      _/home/alex/src/ava-labs/gecko/vms/components/ava       1.079s  coverage: 26.4% of statements
ok      _/home/alex/src/ava-labs/gecko/vms/components/codec     1.797s  coverage: 88.5% of statements
?       _/home/alex/src/ava-labs/gecko/vms/components/core      [no test files]
...
?       _/home/alex/src/ava-labs/gecko/xputtest/chainwallet     [no test files]
?       _/home/alex/src/ava-labs/gecko/xputtest/dagwallet       [no test files]
FAIL
```
It's worth noting that using `GenesisVM` in `setup` also changes the imports of `service_test.go` from 

```
import (
	...
	"github.com/ava-labs/gecko/database/memdb"
	"github.com/ava-labs/gecko/ids"
	"github.com/ava-labs/gecko/snow/engine/common"
	"github.com/ava-labs/gecko/utils/formatting"
	"github.com/ava-labs/gecko/vms/secp256k1fx"
)
```

to

```
import (
	...
	"github.com/ava-labs/gecko/ids"
	"github.com/ava-labs/gecko/utils/formatting"
)

I'll invetigate whether the modules that are no longer imported have some initlialisation that might be getting skipped.

Reply:
If I change the `setup` function to:
```
func setup(t *testing.T) ([]byte, *VM, *Service) {
	genesisBytes, _, vm := GenesisVM(t)
	s := &Service{vm: vm}
	return genesisBytes, vm, s
}
```
The tests pass for me.

When you changed the `setup` function, is it possible that you were grabbing `ctx.Lock` multiple times in the test?

Reply:
> When you changed the `setup` function, is it possible that you were grabbing `ctx.Lock` multiple times in the test?

I don't believe so. Here's a complete transcript of a run I just did. First without the change

```
:~/src/ava-labs/gecko vms-avm-service-test+± git log -n1 | cat
commit 4bf628a8deda618c9d1ca9143d8aa76b26cd8cb6
Merge: b8e1094 7dd42ea
Author: Alex Willmer <alex@moreati.org.uk>
Date:   Mon Apr 13 09:45:26 2020 +0100

    Merge branch 'vms-avm-service-test' of github.com:moreati/gecko into vms-avm-service-test
:~/src/ava-labs/gecko vms-avm-service-test+± echo $GITPATH

:~/src/ava-labs/gecko vms-avm-service-test+± echo $GOPATH
/home/alex/.local/go
:~/src/ava-labs/gecko vms-avm-service-test+± go version
go version go1.13.8 linux/amd64
:~/src/ava-labs/gecko vms-avm-service-test+± scripts/build_test.sh
# github.com/ava-labs/go-ethereum/vendor/gopkg.in/olebedev/go-duktape.v3
In file included from _cgo_export.c:4:
debugger.go:22:13: warning: ‘_duk_debugger_attach’ defined but not used [-Wunused-function]
ok      _/home/alex/src/ava-labs/gecko/api      1.337s  coverage: 63.6% of statements
?       _/home/alex/src/ava-labs/gecko/api/admin        [no test files]
?       _/home/alex/src/ava-labs/gecko/api/ipcs [no test files]
ok      _/home/alex/src/ava-labs/gecko/api/keystore     6.522s  coverage: 77.8% of statements
?       _/home/alex/src/ava-labs/gecko/api/metrics      [no test files]
ok      _/home/alex/src/ava-labs/gecko/cache    1.009s  coverage: 100.0% of statements
?       _/home/alex/src/ava-labs/gecko/chains   [no test files]
ok      _/home/alex/src/ava-labs/gecko/chains/atomic    1.009s  coverage: 97.8% of statements
?       _/home/alex/src/ava-labs/gecko/database [no test files]
ok      _/home/alex/src/ava-labs/gecko/database/encdb   1.011s  coverage: 92.2% of statements
ok      _/home/alex/src/ava-labs/gecko/database/leveldb 4.992s  coverage: 94.6% of statements
ok      _/home/alex/src/ava-labs/gecko/database/memdb   1.008s  coverage: 100.0% of statements
ok      _/home/alex/src/ava-labs/gecko/database/mockdb  1.013s  coverage: 70.3% of statements
?       _/home/alex/src/ava-labs/gecko/database/nodb    [no test files]
ok      _/home/alex/src/ava-labs/gecko/database/prefixdb        1.010s  coverage: 100.0% of statements
ok      _/home/alex/src/ava-labs/gecko/database/versiondb       1.018s  coverage: 98.3% of statements
ok      _/home/alex/src/ava-labs/gecko/genesis  2.281s  coverage: 86.0% of statements
ok      _/home/alex/src/ava-labs/gecko/ids      8.358s  coverage: 79.1% of statements
?       _/home/alex/src/ava-labs/gecko/main     [no test files]
?       _/home/alex/src/ava-labs/gecko/networking       [no test files]
?       _/home/alex/src/ava-labs/gecko/networking/xputtest      [no test files]
?       _/home/alex/src/ava-labs/gecko/node     [no test files]
?       _/home/alex/src/ava-labs/gecko/snow     [no test files]
ok      _/home/alex/src/ava-labs/gecko/snow/choices     1.010s  coverage: 48.4% of statements
ok      _/home/alex/src/ava-labs/gecko/snow/consensus/avalanche 1.024s  coverage: 99.0% of statements
ok      _/home/alex/src/ava-labs/gecko/snow/consensus/snowball  4.561s  coverage: 100.0% of statements
ok      _/home/alex/src/ava-labs/gecko/snow/consensus/snowman   1.981s  coverage: 99.5% of statements
ok      _/home/alex/src/ava-labs/gecko/snow/consensus/snowstorm 4.759s  coverage: 99.5% of statements
ok      _/home/alex/src/ava-labs/gecko/snow/engine/avalanche    1.101s  coverage: 85.9% of statements
?       _/home/alex/src/ava-labs/gecko/snow/engine/avalanche/state      [no test files]
?       _/home/alex/src/ava-labs/gecko/snow/engine/common       [no test files]
ok      _/home/alex/src/ava-labs/gecko/snow/engine/common/queue 1.013s  coverage: 80.1% of statements
ok      _/home/alex/src/ava-labs/gecko/snow/engine/snowman      1.054s  coverage: 82.2% of statements
ok      _/home/alex/src/ava-labs/gecko/snow/events      1.008s  coverage: 73.1% of statements
?       _/home/alex/src/ava-labs/gecko/snow/networking  [no test files]
?       _/home/alex/src/ava-labs/gecko/snow/networking/handler  [no test files]
?       _/home/alex/src/ava-labs/gecko/snow/networking/router   [no test files]
ok      _/home/alex/src/ava-labs/gecko/snow/networking/sender   1.014s  coverage: 12.8% of statements
ok      _/home/alex/src/ava-labs/gecko/snow/networking/timeout  1.059s  coverage: 100.0% of statements
?       _/home/alex/src/ava-labs/gecko/snow/triggers    [no test files]
ok      _/home/alex/src/ava-labs/gecko/snow/validators  1.010s  coverage: 81.6% of statements
ok      _/home/alex/src/ava-labs/gecko/utils    1.016s  coverage: 59.1% of statements
ok      _/home/alex/src/ava-labs/gecko/utils/crypto     8.803s  coverage: 38.3% of statements
ok      _/home/alex/src/ava-labs/gecko/utils/formatting 1.009s  coverage: 93.5% of statements
?       _/home/alex/src/ava-labs/gecko/utils/hashing    [no test files]
?       _/home/alex/src/ava-labs/gecko/utils/json       [no test files]
?       _/home/alex/src/ava-labs/gecko/utils/logging    [no test files]
ok      _/home/alex/src/ava-labs/gecko/utils/math       1.009s  coverage: 100.0% of statements
ok      _/home/alex/src/ava-labs/gecko/utils/random     1.022s  coverage: 95.5% of statements
ok      _/home/alex/src/ava-labs/gecko/utils/timer      1.014s  coverage: 66.7% of statements
?       _/home/alex/src/ava-labs/gecko/utils/units      [no test files]
ok      _/home/alex/src/ava-labs/gecko/utils/wrappers   1.012s  coverage: 61.2% of statements
?       _/home/alex/src/ava-labs/gecko/vms      [no test files]
ok      _/home/alex/src/ava-labs/gecko/vms/avm  1.482s  coverage: 51.9% of statements
ok      _/home/alex/src/ava-labs/gecko/vms/components/ava       1.062s  coverage: 26.4% of statements
ok      _/home/alex/src/ava-labs/gecko/vms/components/codec     1.795s  coverage: 85.1% of statements
?       _/home/alex/src/ava-labs/gecko/vms/components/core      [no test files]
ok      _/home/alex/src/ava-labs/gecko/vms/components/missing   1.009s  coverage: 100.0% of statements
ok      _/home/alex/src/ava-labs/gecko/vms/components/state     1.008s  coverage: 46.5% of statements
ok      _/home/alex/src/ava-labs/gecko/vms/components/verify    1.006s  coverage: 100.0% of statements
ok      _/home/alex/src/ava-labs/gecko/vms/evm  1.112s  coverage: 0.7% of statements
ok      _/home/alex/src/ava-labs/gecko/vms/nftfx        1.077s  coverage: 100.0% of statements
ok      _/home/alex/src/ava-labs/gecko/vms/platformvm   2.709s  coverage: 53.1% of statements
ok      _/home/alex/src/ava-labs/gecko/vms/propertyfx   1.071s  coverage: 100.0% of statements
ok      _/home/alex/src/ava-labs/gecko/vms/secp256k1fx  1.089s  coverage: 96.8% of statements
ok      _/home/alex/src/ava-labs/gecko/vms/spchainvm    2.083s  coverage: 62.4% of statements
ok      _/home/alex/src/ava-labs/gecko/vms/spdagvm      2.139s  coverage: 59.7% of statements
ok      _/home/alex/src/ava-labs/gecko/vms/timestampvm  1.017s  coverage: 59.6% of statements
?       _/home/alex/src/ava-labs/gecko/xputtest [no test files]
ok      _/home/alex/src/ava-labs/gecko/xputtest/avmwallet       1.069s  coverage: 63.9% of statements
?       _/home/alex/src/ava-labs/gecko/xputtest/chainwallet     [no test files]
?       _/home/alex/src/ava-labs/gecko/xputtest/dagwallet       [no test files]
:~/src/ava-labs/gecko vms-avm-service-test+± scripts/build_test.sh
# github.com/ava-labs/go-ethereum/vendor/gopkg.in/olebedev/go-duktape.v3
In file included from _cgo_export.c:4:
debugger.go:22:13: warning: ‘_duk_debugger_attach’ defined but not used [-Wunused-function]
ok      _/home/alex/src/ava-labs/gecko/api      1.011s  coverage: 63.6% of statements
?       _/home/alex/src/ava-labs/gecko/api/admin        [no test files]
?       _/home/alex/src/ava-labs/gecko/api/ipcs [no test files]
ok      _/home/alex/src/ava-labs/gecko/api/keystore     4.898s  coverage: 77.8% of statements
?       _/home/alex/src/ava-labs/gecko/api/metrics      [no test files]
ok      _/home/alex/src/ava-labs/gecko/cache    1.008s  coverage: 100.0% of statements
?       _/home/alex/src/ava-labs/gecko/chains   [no test files]
ok      _/home/alex/src/ava-labs/gecko/chains/atomic    1.008s  coverage: 97.8% of statements
?       _/home/alex/src/ava-labs/gecko/database [no test files]
ok      _/home/alex/src/ava-labs/gecko/database/encdb   1.011s  coverage: 92.2% of statements
ok      _/home/alex/src/ava-labs/gecko/database/leveldb 5.592s  coverage: 94.6% of statements
ok      _/home/alex/src/ava-labs/gecko/database/memdb   1.009s  coverage: 100.0% of statements
ok      _/home/alex/src/ava-labs/gecko/database/mockdb  1.007s  coverage: 70.3% of statements
?       _/home/alex/src/ava-labs/gecko/database/nodb    [no test files]
ok      _/home/alex/src/ava-labs/gecko/database/prefixdb        1.022s  coverage: 100.0% of statements
ok      _/home/alex/src/ava-labs/gecko/database/versiondb       1.038s  coverage: 98.3% of statements
ok      _/home/alex/src/ava-labs/gecko/genesis  1.965s  coverage: 86.0% of statements
ok      _/home/alex/src/ava-labs/gecko/ids      8.201s  coverage: 79.1% of statements
?       _/home/alex/src/ava-labs/gecko/main     [no test files]
?       _/home/alex/src/ava-labs/gecko/networking       [no test files]
?       _/home/alex/src/ava-labs/gecko/networking/xputtest      [no test files]
?       _/home/alex/src/ava-labs/gecko/node     [no test files]
?       _/home/alex/src/ava-labs/gecko/snow     [no test files]
ok      _/home/alex/src/ava-labs/gecko/snow/choices     1.014s  coverage: 48.4% of statements
ok      _/home/alex/src/ava-labs/gecko/snow/consensus/avalanche 1.035s  coverage: 99.0% of statements
ok      _/home/alex/src/ava-labs/gecko/snow/consensus/snowball  4.140s  coverage: 100.0% of statements
ok      _/home/alex/src/ava-labs/gecko/snow/consensus/snowman   1.744s  coverage: 99.5% of statements
ok      _/home/alex/src/ava-labs/gecko/snow/consensus/snowstorm 4.349s  coverage: 99.5% of statements
ok      _/home/alex/src/ava-labs/gecko/snow/engine/avalanche    1.133s  coverage: 85.9% of statements
?       _/home/alex/src/ava-labs/gecko/snow/engine/avalanche/state      [no test files]
?       _/home/alex/src/ava-labs/gecko/snow/engine/common       [no test files]
ok      _/home/alex/src/ava-labs/gecko/snow/engine/common/queue 1.011s  coverage: 80.1% of statements
ok      _/home/alex/src/ava-labs/gecko/snow/engine/snowman      1.064s  coverage: 82.2% of statements
ok      _/home/alex/src/ava-labs/gecko/snow/events      1.007s  coverage: 73.1% of statements
?       _/home/alex/src/ava-labs/gecko/snow/networking  [no test files]
?       _/home/alex/src/ava-labs/gecko/snow/networking/handler  [no test files]
?       _/home/alex/src/ava-labs/gecko/snow/networking/router   [no test files]
ok      _/home/alex/src/ava-labs/gecko/snow/networking/sender   1.010s  coverage: 12.8% of statements
ok      _/home/alex/src/ava-labs/gecko/snow/networking/timeout  1.059s  coverage: 100.0% of statements
?       _/home/alex/src/ava-labs/gecko/snow/triggers    [no test files]
ok      _/home/alex/src/ava-labs/gecko/snow/validators  1.008s  coverage: 81.6% of statements
ok      _/home/alex/src/ava-labs/gecko/utils    1.012s  coverage: 59.1% of statements
ok      _/home/alex/src/ava-labs/gecko/utils/crypto     11.161s coverage: 38.3% of statements
ok      _/home/alex/src/ava-labs/gecko/utils/formatting 1.008s  coverage: 93.5% of statements
?       _/home/alex/src/ava-labs/gecko/utils/hashing    [no test files]
?       _/home/alex/src/ava-labs/gecko/utils/json       [no test files]
?       _/home/alex/src/ava-labs/gecko/utils/logging    [no test files]
ok      _/home/alex/src/ava-labs/gecko/utils/math       1.008s  coverage: 100.0% of statements
ok      _/home/alex/src/ava-labs/gecko/utils/random     1.032s  coverage: 95.5% of statements
ok      _/home/alex/src/ava-labs/gecko/utils/timer      1.014s  coverage: 66.7% of statements
?       _/home/alex/src/ava-labs/gecko/utils/units      [no test files]
ok      _/home/alex/src/ava-labs/gecko/utils/wrappers   1.010s  coverage: 61.2% of statements
?       _/home/alex/src/ava-labs/gecko/vms      [no test files]
ok      _/home/alex/src/ava-labs/gecko/vms/avm  1.432s  coverage: 51.9% of statements
ok      _/home/alex/src/ava-labs/gecko/vms/components/ava       1.101s  coverage: 26.4% of statements
ok      _/home/alex/src/ava-labs/gecko/vms/components/codec     1.847s  coverage: 85.1% of statements
?       _/home/alex/src/ava-labs/gecko/vms/components/core      [no test files]
ok      _/home/alex/src/ava-labs/gecko/vms/components/missing   1.012s  coverage: 100.0% of statements
ok      _/home/alex/src/ava-labs/gecko/vms/components/state     1.008s  coverage: 46.5% of statements
ok      _/home/alex/src/ava-labs/gecko/vms/components/verify    1.019s  coverage: 100.0% of statements
ok      _/home/alex/src/ava-labs/gecko/vms/evm  1.114s  coverage: 0.7% of statements
ok      _/home/alex/src/ava-labs/gecko/vms/nftfx        1.065s  coverage: 100.0% of statements
ok      _/home/alex/src/ava-labs/gecko/vms/platformvm   2.882s  coverage: 53.1% of statements
ok      _/home/alex/src/ava-labs/gecko/vms/propertyfx   1.064s  coverage: 100.0% of statements
ok      _/home/alex/src/ava-labs/gecko/vms/secp256k1fx  1.105s  coverage: 96.8% of statements
ok      _/home/alex/src/ava-labs/gecko/vms/spchainvm    2.123s  coverage: 62.4% of statements
ok      _/home/alex/src/ava-labs/gecko/vms/spdagvm      2.139s  coverage: 59.7% of statements
ok      _/home/alex/src/ava-labs/gecko/vms/timestampvm  1.015s  coverage: 59.6% of statements
?       _/home/alex/src/ava-labs/gecko/xputtest [no test files]
ok      _/home/alex/src/ava-labs/gecko/xputtest/avmwallet       1.069s  coverage: 63.9% of statements
?       _/home/alex/src/ava-labs/gecko/xputtest/chainwallet     [no test files]
?       _/home/alex/src/ava-labs/gecko/xputtest/dagwallet       [no test files]
```

then with the change

```
:~/src/ava-labs/gecko vms-avm-service-test+± git stash pop
On branch vms-avm-service-test
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

        modified:   scripts/build_test.sh
        modified:   vms/avm/service_test.go

no changes added to commit (use "git add" and/or "git commit -a")
Dropped refs/stash@{0} (b084bad4e845809b277227dfd0ff41b70266ac12)
:~/src/ava-labs/gecko vms-avm-service-test(+3/-25)+± git diff | cat
diff --git a/scripts/build_test.sh b/scripts/build_test.sh
index f627613..12cb56a 100755
--- a/scripts/build_test.sh
+++ b/scripts/build_test.sh
@@ -5,4 +5,4 @@
 SRC_DIR="$(dirname "${BASH_SOURCE[0]}")"
 source "$SRC_DIR/env.sh"
 
-go test -race -coverprofile=coverage.out -covermode=atomic ./...
+go test --timeout=30s -race -coverprofile=coverage.out -covermode=atomic ./...
diff --git a/vms/avm/service_test.go b/vms/avm/service_test.go
index 918f35e..70adc74 100644
--- a/vms/avm/service_test.go
+++ b/vms/avm/service_test.go
@@ -9,35 +9,13 @@ import (
 
        "github.com/ava-labs/gecko/snow/choices"
 
-       "github.com/ava-labs/gecko/database/memdb"
        "github.com/ava-labs/gecko/ids"
-       "github.com/ava-labs/gecko/snow/engine/common"
        "github.com/ava-labs/gecko/utils/formatting"
-       "github.com/ava-labs/gecko/vms/secp256k1fx"
 )
 
+//foo
 func setup(t *testing.T) ([]byte, *VM, *Service) {
-       genesisBytes := BuildGenesisTest(t)
-
-       ctx.Lock.Lock()
-
-       // This is VM initilialzation is very silimar to that done by GenesisVM().
-       // However that function also sets vm.batchTimeout = 0, which causes the
-       // tests in this module to deadlock.
-       vm := &VM{}
-       err := vm.Initialize(
-               ctx,
-               memdb.New(),
-               genesisBytes,
-               make(chan common.Message, 1),
-               []*common.Fx{&common.Fx{
-                       ID: ids.Empty,
-                       Fx: &secp256k1fx.Fx{},
-               }},
-       )
-       if err != nil {
-               t.Fatal(err)
-       }
+       genesisBytes, _, vm := GenesisVM(t)
        s := &Service{vm: vm}
        return genesisBytes, vm, s
 }
:~/src/ava-labs/gecko vms-avm-service-test(+3/-25)+± scripts/build_test.sh
ok      _/home/alex/src/ava-labs/gecko/api      1.010s  coverage: 63.6% of statements
?       _/home/alex/src/ava-labs/gecko/api/admin        [no test files]
?       _/home/alex/src/ava-labs/gecko/api/ipcs [no test files]
# github.com/ava-labs/go-ethereum/vendor/gopkg.in/olebedev/go-duktape.v3
In file included from _cgo_export.c:4:
debugger.go:22:13: warning: ‘_duk_debugger_attach’ defined but not used [-Wunused-function]
ok      _/home/alex/src/ava-labs/gecko/api/keystore     3.961s  coverage: 77.8% of statements
?       _/home/alex/src/ava-labs/gecko/api/metrics      [no test files]
ok      _/home/alex/src/ava-labs/gecko/cache    1.007s  coverage: 100.0% of statements
?       _/home/alex/src/ava-labs/gecko/chains   [no test files]
ok      _/home/alex/src/ava-labs/gecko/chains/atomic    1.008s  coverage: 97.8% of statements
?       _/home/alex/src/ava-labs/gecko/database [no test files]
ok      _/home/alex/src/ava-labs/gecko/database/encdb   1.012s  coverage: 92.2% of statements
ok      _/home/alex/src/ava-labs/gecko/database/leveldb 3.249s  coverage: 94.6% of statements
ok      _/home/alex/src/ava-labs/gecko/database/memdb   1.007s  coverage: 100.0% of statements
ok      _/home/alex/src/ava-labs/gecko/database/mockdb  1.007s  coverage: 70.3% of statements
?       _/home/alex/src/ava-labs/gecko/database/nodb    [no test files]
ok      _/home/alex/src/ava-labs/gecko/database/prefixdb        1.018s  coverage: 100.0% of statements
ok      _/home/alex/src/ava-labs/gecko/database/versiondb       1.040s  coverage: 98.3% of statements
ok      _/home/alex/src/ava-labs/gecko/genesis  1.931s  coverage: 86.0% of statements
ok      _/home/alex/src/ava-labs/gecko/ids      8.020s  coverage: 79.1% of statements
?       _/home/alex/src/ava-labs/gecko/main     [no test files]
?       _/home/alex/src/ava-labs/gecko/networking       [no test files]
?       _/home/alex/src/ava-labs/gecko/networking/xputtest      [no test files]
?       _/home/alex/src/ava-labs/gecko/node     [no test files]
?       _/home/alex/src/ava-labs/gecko/snow     [no test files]
ok      _/home/alex/src/ava-labs/gecko/snow/choices     1.039s  coverage: 48.4% of statements
ok      _/home/alex/src/ava-labs/gecko/snow/consensus/avalanche 1.030s  coverage: 99.0% of statements
ok      _/home/alex/src/ava-labs/gecko/snow/consensus/snowball  3.906s  coverage: 100.0% of statements
ok      _/home/alex/src/ava-labs/gecko/snow/consensus/snowman   1.753s  coverage: 99.5% of statements
ok      _/home/alex/src/ava-labs/gecko/snow/consensus/snowstorm 4.316s  coverage: 99.5% of statements
ok      _/home/alex/src/ava-labs/gecko/snow/engine/avalanche    1.116s  coverage: 85.9% of statements
?       _/home/alex/src/ava-labs/gecko/snow/engine/avalanche/state      [no test files]
?       _/home/alex/src/ava-labs/gecko/snow/engine/common       [no test files]
ok      _/home/alex/src/ava-labs/gecko/snow/engine/common/queue 1.010s  coverage: 80.1% of statements
ok      _/home/alex/src/ava-labs/gecko/snow/engine/snowman      1.055s  coverage: 82.2% of statements
ok      _/home/alex/src/ava-labs/gecko/snow/events      1.006s  coverage: 73.1% of statements
?       _/home/alex/src/ava-labs/gecko/snow/networking  [no test files]
?       _/home/alex/src/ava-labs/gecko/snow/networking/handler  [no test files]
?       _/home/alex/src/ava-labs/gecko/snow/networking/router   [no test files]
ok      _/home/alex/src/ava-labs/gecko/snow/networking/sender   1.015s  coverage: 12.8% of statements
ok      _/home/alex/src/ava-labs/gecko/snow/networking/timeout  1.059s  coverage: 100.0% of statements
?       _/home/alex/src/ava-labs/gecko/snow/triggers    [no test files]
ok      _/home/alex/src/ava-labs/gecko/snow/validators  1.008s  coverage: 81.6% of statements
ok      _/home/alex/src/ava-labs/gecko/utils    1.012s  coverage: 59.1% of statements
ok      _/home/alex/src/ava-labs/gecko/utils/crypto     6.863s  coverage: 38.3% of statements
ok      _/home/alex/src/ava-labs/gecko/utils/formatting 1.009s  coverage: 93.5% of statements
?       _/home/alex/src/ava-labs/gecko/utils/hashing    [no test files]
?       _/home/alex/src/ava-labs/gecko/utils/json       [no test files]
?       _/home/alex/src/ava-labs/gecko/utils/logging    [no test files]
ok      _/home/alex/src/ava-labs/gecko/utils/math       1.010s  coverage: 100.0% of statements
ok      _/home/alex/src/ava-labs/gecko/utils/random     1.026s  coverage: 95.5% of statements
ok      _/home/alex/src/ava-labs/gecko/utils/timer      1.014s  coverage: 66.7% of statements
?       _/home/alex/src/ava-labs/gecko/utils/units      [no test files]
ok      _/home/alex/src/ava-labs/gecko/utils/wrappers   1.012s  coverage: 61.2% of statements
?       _/home/alex/src/ava-labs/gecko/vms      [no test files]
coverage: 46.1% of statements
panic: test timed out after 30s

goroutine 140 [running]:
testing.(*M).startAlarm.func1()
        /usr/local/go/src/testing/testing.go:1377 +0x11c
created by time.goFunc
        /usr/local/go/src/time/sleep.go:168 +0x52

goroutine 1 [chan receive]:
testing.(*T).Run(0xc00017e000, 0xc2e39e, 0x17, 0xc48a50, 0x1)
        /usr/local/go/src/testing/testing.go:961 +0x68a
testing.runTests.func1(0xc00017e000)
        /usr/local/go/src/testing/testing.go:1202 +0xa7
testing.tRunner(0xc00017e000, 0xc0000c5cd8)
        /usr/local/go/src/testing/testing.go:909 +0x19a
testing.runTests(0xc000132200, 0x107d220, 0x48, 0x48, 0x0)
        /usr/local/go/src/testing/testing.go:1200 +0x522
testing.(*M).Run(0xc000178000, 0x0)
        /usr/local/go/src/testing/testing.go:1117 +0x300
main.main()
        _testmain.go:268 +0x338

goroutine 82 [select]:
github.com/ava-labs/gecko/utils/timer.(*Timer).Dispatch(0xc00049e960)
        /home/alex/.local/go/src/github.com/ava-labs/gecko/utils/timer/timer.go:90 +0x289
github.com/ava-labs/gecko/utils/logging.NoLog.RecoverAndPanic(...)
        /home/alex/.local/go/src/github.com/ava-labs/gecko/utils/logging/test_log.go:53
created by _/home/alex/src/ava-labs/gecko/vms/avm.(*VM).Initialize
        /home/alex/src/ava-labs/gecko/vms/avm/vm.go:193 +0x1a84

goroutine 49 [select]:
github.com/ava-labs/gecko/utils/timer.(*Timer).Dispatch(0xc0001feb10)
        /home/alex/.local/go/src/github.com/ava-labs/gecko/utils/timer/timer.go:90 +0x289
github.com/ava-labs/gecko/utils/logging.NoLog.RecoverAndPanic(...)
        /home/alex/.local/go/src/github.com/ava-labs/gecko/utils/logging/test_log.go:53
created by _/home/alex/src/ava-labs/gecko/vms/avm.(*VM).Initialize
        /home/alex/src/ava-labs/gecko/vms/avm/vm.go:193 +0x1a84

goroutine 51 [select]:
github.com/ava-labs/gecko/utils/timer.(*Timer).Dispatch(0xc00028c360)
        /home/alex/.local/go/src/github.com/ava-labs/gecko/utils/timer/timer.go:90 +0x289
github.com/ava-labs/gecko/utils/logging.NoLog.RecoverAndPanic(...)
        /home/alex/.local/go/src/github.com/ava-labs/gecko/utils/logging/test_log.go:53
created by _/home/alex/src/ava-labs/gecko/vms/avm.(*VM).Initialize
        /home/alex/src/ava-labs/gecko/vms/avm/vm.go:193 +0x1a84

goroutine 53 [select]:
github.com/ava-labs/gecko/utils/timer.(*Timer).Dispatch(0xc00028da10)
        /home/alex/.local/go/src/github.com/ava-labs/gecko/utils/timer/timer.go:90 +0x289
github.com/ava-labs/gecko/utils/logging.NoLog.RecoverAndPanic(...)
        /home/alex/.local/go/src/github.com/ava-labs/gecko/utils/logging/test_log.go:53
created by _/home/alex/src/ava-labs/gecko/vms/avm.(*VM).Initialize
        /home/alex/src/ava-labs/gecko/vms/avm/vm.go:193 +0x1a84

goroutine 55 [select]:
github.com/ava-labs/gecko/utils/timer.(*Timer).Dispatch(0xc000337110)
        /home/alex/.local/go/src/github.com/ava-labs/gecko/utils/timer/timer.go:90 +0x289
github.com/ava-labs/gecko/utils/logging.NoLog.RecoverAndPanic(...)
        /home/alex/.local/go/src/github.com/ava-labs/gecko/utils/logging/test_log.go:53
created by _/home/alex/src/ava-labs/gecko/vms/avm.(*VM).Initialize
        /home/alex/src/ava-labs/gecko/vms/avm/vm.go:193 +0x1a84

goroutine 57 [select]:
github.com/ava-labs/gecko/utils/timer.(*Timer).Dispatch(0xc0003fe870)
        /home/alex/.local/go/src/github.com/ava-labs/gecko/utils/timer/timer.go:90 +0x289
github.com/ava-labs/gecko/utils/logging.NoLog.RecoverAndPanic(...)
        /home/alex/.local/go/src/github.com/ava-labs/gecko/utils/logging/test_log.go:53
created by _/home/alex/src/ava-labs/gecko/vms/avm.(*VM).Initialize
        /home/alex/src/ava-labs/gecko/vms/avm/vm.go:193 +0x1a84

goroutine 62 [select]:
github.com/ava-labs/gecko/utils/timer.(*Timer).Dispatch(0xc0004c0e10)
        /home/alex/.local/go/src/github.com/ava-labs/gecko/utils/timer/timer.go:90 +0x289
github.com/ava-labs/gecko/utils/logging.NoLog.RecoverAndPanic(...)
        /home/alex/.local/go/src/github.com/ava-labs/gecko/utils/logging/test_log.go:53
created by _/home/alex/src/ava-labs/gecko/vms/avm.(*VM).Initialize
        /home/alex/src/ava-labs/gecko/vms/avm/vm.go:193 +0x1a84

goroutine 64 [select]:
github.com/ava-labs/gecko/utils/timer.(*Timer).Dispatch(0xc0001489c0)
        /home/alex/.local/go/src/github.com/ava-labs/gecko/utils/timer/timer.go:90 +0x289
github.com/ava-labs/gecko/utils/logging.NoLog.RecoverAndPanic(...)
        /home/alex/.local/go/src/github.com/ava-labs/gecko/utils/logging/test_log.go:53
created by _/home/alex/src/ava-labs/gecko/vms/avm.(*VM).Initialize
        /home/alex/src/ava-labs/gecko/vms/avm/vm.go:193 +0x1a84

goroutine 7 [select]:
github.com/ava-labs/gecko/utils/timer.(*Timer).Dispatch(0xc00028db00)
        /home/alex/.local/go/src/github.com/ava-labs/gecko/utils/timer/timer.go:90 +0x289
github.com/ava-labs/gecko/utils/logging.NoLog.RecoverAndPanic(...)
        /home/alex/.local/go/src/github.com/ava-labs/gecko/utils/logging/test_log.go:53
created by _/home/alex/src/ava-labs/gecko/vms/avm.(*VM).Initialize
        /home/alex/src/ava-labs/gecko/vms/avm/vm.go:193 +0x1a84

goroutine 8 [select]:
github.com/ava-labs/gecko/utils/timer.(*Timer).Dispatch(0xc00049fdd0)
        /home/alex/.local/go/src/github.com/ava-labs/gecko/utils/timer/timer.go:90 +0x289
github.com/ava-labs/gecko/utils/logging.NoLog.RecoverAndPanic(...)
        /home/alex/.local/go/src/github.com/ava-labs/gecko/utils/logging/test_log.go:53
created by _/home/alex/src/ava-labs/gecko/vms/avm.(*VM).Initialize
        /home/alex/src/ava-labs/gecko/vms/avm/vm.go:193 +0x1a84

goroutine 10 [select]:
github.com/ava-labs/gecko/utils/timer.(*Timer).Dispatch(0xc000148f00)
        /home/alex/.local/go/src/github.com/ava-labs/gecko/utils/timer/timer.go:90 +0x289
github.com/ava-labs/gecko/utils/logging.NoLog.RecoverAndPanic(...)
        /home/alex/.local/go/src/github.com/ava-labs/gecko/utils/logging/test_log.go:53
created by _/home/alex/src/ava-labs/gecko/vms/avm.(*VM).Initialize
        /home/alex/src/ava-labs/gecko/vms/avm/vm.go:193 +0x1a84

goroutine 14 [select]:
github.com/ava-labs/gecko/utils/timer.(*Timer).Dispatch(0xc0003ffc20)
        /home/alex/.local/go/src/github.com/ava-labs/gecko/utils/timer/timer.go:90 +0x289
github.com/ava-labs/gecko/utils/logging.NoLog.RecoverAndPanic(...)
        /home/alex/.local/go/src/github.com/ava-labs/gecko/utils/logging/test_log.go:53
created by _/home/alex/src/ava-labs/gecko/vms/avm.(*VM).Initialize
        /home/alex/src/ava-labs/gecko/vms/avm/vm.go:193 +0x1a84

goroutine 15 [select]:
github.com/ava-labs/gecko/utils/timer.(*Timer).Dispatch(0xc00049eab0)
        /home/alex/.local/go/src/github.com/ava-labs/gecko/utils/timer/timer.go:90 +0x289
github.com/ava-labs/gecko/utils/logging.NoLog.RecoverAndPanic(...)
        /home/alex/.local/go/src/github.com/ava-labs/gecko/utils/logging/test_log.go:53
created by _/home/alex/src/ava-labs/gecko/vms/avm.(*VM).Initialize
        /home/alex/src/ava-labs/gecko/vms/avm/vm.go:193 +0x1a84

goroutine 115 [select]:
github.com/ava-labs/gecko/utils/timer.(*Timer).Dispatch(0xc00015ecc0)
        /home/alex/.local/go/src/github.com/ava-labs/gecko/utils/timer/timer.go:90 +0x289
github.com/ava-labs/gecko/utils/logging.NoLog.RecoverAndPanic(...)
        /home/alex/.local/go/src/github.com/ava-labs/gecko/utils/logging/test_log.go:53
created by _/home/alex/src/ava-labs/gecko/vms/avm.(*VM).Initialize
        /home/alex/src/ava-labs/gecko/vms/avm/vm.go:193 +0x1a84

goroutine 117 [select]:
github.com/ava-labs/gecko/utils/timer.(*Timer).Dispatch(0xc000153290)
        /home/alex/.local/go/src/github.com/ava-labs/gecko/utils/timer/timer.go:90 +0x289
github.com/ava-labs/gecko/utils/logging.NoLog.RecoverAndPanic(...)
        /home/alex/.local/go/src/github.com/ava-labs/gecko/utils/logging/test_log.go:53
created by _/home/alex/src/ava-labs/gecko/vms/avm.(*VM).Initialize
        /home/alex/src/ava-labs/gecko/vms/avm/vm.go:193 +0x1a84

goroutine 130 [select]:
github.com/ava-labs/gecko/utils/timer.(*Timer).Dispatch(0xc0001997d0)
        /home/alex/.local/go/src/github.com/ava-labs/gecko/utils/timer/timer.go:90 +0x289
github.com/ava-labs/gecko/utils/logging.NoLog.RecoverAndPanic(...)
        /home/alex/.local/go/src/github.com/ava-labs/gecko/utils/logging/test_log.go:53
created by _/home/alex/src/ava-labs/gecko/vms/avm.(*VM).Initialize
        /home/alex/src/ava-labs/gecko/vms/avm/vm.go:193 +0x1a84

goroutine 132 [select]:
github.com/ava-labs/gecko/utils/timer.(*Timer).Dispatch(0xc000199b90)
        /home/alex/.local/go/src/github.com/ava-labs/gecko/utils/timer/timer.go:90 +0x289
github.com/ava-labs/gecko/utils/logging.NoLog.RecoverAndPanic(...)
        /home/alex/.local/go/src/github.com/ava-labs/gecko/utils/logging/test_log.go:53
created by _/home/alex/src/ava-labs/gecko/vms/avm.(*VM).Initialize
        /home/alex/src/ava-labs/gecko/vms/avm/vm.go:193 +0x1a84

goroutine 134 [select]:
github.com/ava-labs/gecko/utils/timer.(*Timer).Dispatch(0xc0000b1b30)
        /home/alex/.local/go/src/github.com/ava-labs/gecko/utils/timer/timer.go:90 +0x289
github.com/ava-labs/gecko/utils/logging.NoLog.RecoverAndPanic(...)
        /home/alex/.local/go/src/github.com/ava-labs/gecko/utils/logging/test_log.go:53
created by _/home/alex/src/ava-labs/gecko/vms/avm.(*VM).Initialize
        /home/alex/src/ava-labs/gecko/vms/avm/vm.go:193 +0x1a84

goroutine 138 [semacquire]:
sync.runtime_Semacquire(0xc000149130)
        /usr/local/go/src/runtime/sema.go:56 +0x42
sync.(*WaitGroup).Wait(0xc000149128)
        /usr/local/go/src/sync/waitgroup.go:130 +0xb1
github.com/ava-labs/gecko/utils/timer.(*Timer).Stop(0xc000149110)
        /home/alex/.local/go/src/github.com/ava-labs/gecko/utils/timer/timer.go:65 +0xe2
_/home/alex/src/ava-labs/gecko/vms/avm.(*VM).Shutdown(0xc000138ea0)
        /home/alex/src/ava-labs/gecko/vms/avm/vm.go:201 +0x6d
_/home/alex/src/ava-labs/gecko/vms/avm.TestCreateFixedCapAsset(0xc00017e200)
        /home/alex/src/ava-labs/gecko/vms/avm/service_test.go:237 +0x494
testing.tRunner(0xc00017e200, 0xc48a50)
        /usr/local/go/src/testing/testing.go:909 +0x19a
created by testing.(*T).Run
        /usr/local/go/src/testing/testing.go:960 +0x652

goroutine 139 [semacquire]:
sync.runtime_SemacquireMutex(0xc00014a03c, 0x900000000, 0x1)
        /usr/local/go/src/runtime/sema.go:71 +0x47
sync.(*Mutex).lockSlow(0xc00014a038)
        /usr/local/go/src/sync/mutex.go:138 +0x1c1
sync.(*Mutex).Lock(0xc00014a038)
        /usr/local/go/src/sync/mutex.go:81 +0x7d
sync.(*RWMutex).Lock(0xc00014a038)
        /usr/local/go/src/sync/rwmutex.go:98 +0x4a
_/home/alex/src/ava-labs/gecko/vms/avm.(*VM).Initialize.func1()
        /home/alex/src/ava-labs/gecko/vms/avm/vm.go:188 +0x7c
github.com/ava-labs/gecko/utils/timer.(*Timer).Dispatch(0xc000149110)
        /home/alex/.local/go/src/github.com/ava-labs/gecko/utils/timer/timer.go:83 +0x195
github.com/ava-labs/gecko/utils/logging.NoLog.RecoverAndPanic(...)
        /home/alex/.local/go/src/github.com/ava-labs/gecko/utils/logging/test_log.go:53
created by _/home/alex/src/ava-labs/gecko/vms/avm.(*VM).Initialize
        /home/alex/src/ava-labs/gecko/vms/avm/vm.go:193 +0x1a84
FAIL    _/home/alex/src/ava-labs/gecko/vms/avm  30.094s
ok      _/home/alex/src/ava-labs/gecko/vms/components/ava       1.064s  coverage: 26.4% of statements
ok      _/home/alex/src/ava-labs/gecko/vms/components/codec     1.758s  coverage: 85.1% of statements
?       _/home/alex/src/ava-labs/gecko/vms/components/core      [no test files]
ok      _/home/alex/src/ava-labs/gecko/vms/components/missing   1.009s  coverage: 100.0% of statements
ok      _/home/alex/src/ava-labs/gecko/vms/components/state     1.012s  coverage: 46.5% of statements
ok      _/home/alex/src/ava-labs/gecko/vms/components/verify    1.008s  coverage: 100.0% of statements
ok      _/home/alex/src/ava-labs/gecko/vms/evm  1.113s  coverage: 0.7% of statements
ok      _/home/alex/src/ava-labs/gecko/vms/nftfx        1.063s  coverage: 100.0% of statements
ok      _/home/alex/src/ava-labs/gecko/vms/platformvm   2.773s  coverage: 53.1% of statements
ok      _/home/alex/src/ava-labs/gecko/vms/propertyfx   1.065s  coverage: 100.0% of statements
ok      _/home/alex/src/ava-labs/gecko/vms/secp256k1fx  1.079s  coverage: 96.8% of statements
ok      _/home/alex/src/ava-labs/gecko/vms/spchainvm    2.082s  coverage: 62.4% of statements
ok      _/home/alex/src/ava-labs/gecko/vms/spdagvm      2.146s  coverage: 59.7% of statements
ok      _/home/alex/src/ava-labs/gecko/vms/timestampvm  1.012s  coverage: 59.6% of statements
?       _/home/alex/src/ava-labs/gecko/xputtest [no test files]
ok      _/home/alex/src/ava-labs/gecko/xputtest/avmwallet       1.068s  coverage: 63.9% of statements
?       _/home/alex/src/ava-labs/gecko/xputtest/chainwallet     [no test files]
?       _/home/alex/src/ava-labs/gecko/xputtest/dagwallet       [no test files]
FAIL
:~/src/ava-labs/gecko vms-avm-service-test(+3/-25)+ 1± 
```

Reply:
You're right. I think this is a bug. I'll look into this more tomorrow.

Reply:
Please see #66.

Reply:
This PR is large enough already. I'll take a stab at #66 seperately